### PR TITLE
Deserialize: Commit changes before activation (2)

### PR DESCRIPTION
**Pull request by mbtools**
Follow-up #4509, #4518

Check program type before calling RPY_PROGRAM_INSERT

### CODE REVIEW

**Comment by larshp** [comment]
before we had 2 different INSERT REPORT, now we have 3 different paths for INSERT REPORT

what is solved or not handled in the different scenarios?

**Comment by mbtools** [comment]
I refactored `METHOD deserialize_program` and it should be much clearer now: 
first AG tries the standard `RPY_INSERT_PROGRAM` function and in case of errors we do direct `INSERT REPORT`.

I tested it with various PROG and FUGR cases from https://github.com/abapGit-tests and it works fine.

**Comment by larshp** [comment]
awaiting review from @larshp

**Comment by larshp** [comment]
the extracting to new methods is good, but I have problems keeping track when there are also logical and critical changes mixed in, PROGs are a crucial part

in method `insert_program`, there are now 3 code paths instead of 2 before, I'm not sure what this will change? If its the commits, then the new code will have the same problem as the old, as the 2 old paths still exist unchanged?

**Review by larshp — CHANGES_REQUESTED** [review]
see comment above

**Comment by mbtools** [comment]
The code is hard to follow as diffs. It's much better in the system. It's core, so yes, this should fail. 

In order for PROG to work properly, you need to `INSERT REPORT` with both 'A' and 'I" state. That's what the RPY function does. Since it's standard, it goes first and is safer than what we previously had (path 1). If it fails, then the hardcoded `INSERT REPORT` options for CI variants (path 2) or FUGR includes (path 3) come into play. 

I'm setting up another system to run the AG/CI tests again. We can put this on hold until then.

**Comment by larshp** [comment]
does it solve any reproduceable issue?

The FM is probably safer yea, but maintenance wise, I think its better to have two branches executed often, than 3 branches where one is a special rarely used branch.

**Comment by mbtools** [comment]
The reproducible case is in #4500. In case of syntax errors after a pull, you can not edit the inactive code of PROGs (program not found in SE38).

We have three cases:
1. regular programs (rpy branch)
2. fugr, logical dbs, etc (things that rpy doens't handle)
3. things that require `extension type`

The rare `extention type` branch shouldn't be part of `deserialize_program` since that's are not executable code (#584). We could move it to `zcl_abapgit_object_prog`. That would make `deserialize_program` simpler: rpy or insert report directly.

**Comment by mbtools** [comment]
replaced by #4657
